cpp/tests/structure/mg_select_random_vertices_test.cpp

```diff
@@ -0,0 +1,240 @@
+  Tests_MGSelectRandomVertices() {}
+
+  static void SetUpTestCase() { handle_ = cugraph::test::initialize_mg_handle(); }
+  static void TearDownTestCase() { handle_.reset(); }
+
+  virtual void SetUp() {}
+  virtual void TearDown() {}
+
+  template <typename vertex_t, typename edge_t, typename weight_t>
+  void run_current_test(std::tuple<SelectRandomVertices_Usecase, input_usecase_t> const& param)
+  {
+    auto [select_random_vertices_usecase, input_usecase] = param;
+
+    auto const comm_rank = handle_->get_comms().get_rank();
+    auto const comm_size = handle_->get_comms().get_size();
+
+    HighResTimer hr_timer{};
+
+    if (cugraph::test::g_perf) {
+      RAFT_CUDA_TRY(cudaDeviceSynchronize());
+      handle_->get_comms().barrier();
+      hr_timer.start("MG Construct graph");
+    }
+
+    auto [mg_graph, mg_edge_weights, mg_renumber_map] =
+      cugraph::test::construct_graph<vertex_t, edge_t, weight_t, false, true>(
+        *handle_, input_usecase, false, true);
+
+    if (cugraph::test::g_perf) {
+      RAFT_CUDA_TRY(cudaDeviceSynchronize());
+      handle_->get_comms().barrier();
+      hr_timer.stop();
+      hr_timer.display_and_clear(std::cout);
+    }
+
+    auto mg_graph_view = mg_graph.view();
+    auto mg_edge_weight_view =
+      mg_edge_weights ? std::make_optional((*mg_edge_weights).view()) : std::nullopt;
+
+    raft::random::RngState rng_state(comm_rank);
+
+    //
+    // Test sampling from a distributed set
+    //
+
+    {
+      // Generate distributed vertex set to sample from
+      srand((unsigned)time(NULL));
+      std::vector<vertex_t> h_given_set(rand() % mg_graph_view.local_vertex_partition_range_size() +
+                                        1);
+
+      for (int i = 0; i < h_given_set.size(); i++) {
+        h_given_set[i] = mg_graph_view.local_vertex_partition_range_first() +
+                         rand() % mg_graph_view.local_vertex_partition_range_size();
+      }
+
+      std::sort(h_given_set.begin(), h_given_set.end());
+      auto last = std::unique(h_given_set.begin(), h_given_set.end());
+      h_given_set.erase(last, h_given_set.end());
+
+      // Compute size of the distributed vertex set
+      int num_of_given_set = static_cast<int>(h_given_set.size());
+
+      rmm::device_uvector<int> d_num_of_given_set(1, handle_->get_stream());
+      raft::update_device(d_num_of_given_set.data(), &num_of_given_set, 1, handle_->get_stream());
+      handle_->get_comms().allreduce(d_num_of_given_set.data(),
+                                     d_num_of_given_set.data(),
+                                     1,
+                                     raft::comms::op_t::SUM,
+                                     handle_->get_stream());
+      raft::update_host(&num_of_given_set, d_num_of_given_set.data(), 1, handle_->get_stream());
+      auto status = handle_->get_comms().sync_stream(handle_->get_stream());
+      CUGRAPH_EXPECTS(status == raft::comms::status_t::SUCCESS, "sync_stream() failure.");
```

Comment:
https://github.com/rapidsai/cugraph/blob/branch-23.06/cpp/include/cugraph/utilities/host_scalar_comm.hpp#L126

Reply:
I haven't found any test code that uses  `host_scalar_allreduce` in a cpp file. If we use so, the the file has to be a .cu file.  Am I getting it wrong?

Reply:
If the header file is .hpp, it can be included in both cpp and cu. If the header file is .cuh, it can only be compiled with nvcc. If you can't call host_scalar_allreduce from .cpp, cugraph/utilities/host_scalar_comm.hpp should not have .hpp extension. Let me know if you can't call this from .cpp file (I don't think this will be a case, but just in case... roughly speaking, you need nvcc if you need to call device functions or run thrust functions with device execution policy, I don't think this is a case for host_scalar_comm.hpp).